### PR TITLE
✨ Auto post matches and match highlights into db

### PR DESCRIPTION
**Pull request by webdevelukas**
Should solve isssue #18 

## Important
To make it work correctly you need to delete the debug urls in `dist/overwolf/manifest.json`

## Done
[X] Add functionality to auto post new match into db
[X] Add functionality to auto post new highlights into db

## ToDo:
[ ] Get username of current user
[ ] Make sure that the same match isn't posted more than once

## Open issues / questions
#40

### CODE REVIEW

File: src/server.ts [hunk 1]
```diff
@@ -21,6 +22,8 @@ if (typeof MONGODB_URI !== 'string') {
 const app = express();
 app.use(helmet());
 
+app.use(cors());
```
Comment (lmachens) [inline src/server.ts]: Let's discuss how we want to handle cross-origin requests.
This will allow requests from any source.
I would propose to limit requests from the Overwolf app only (and localhost in development).
This would be possible by adding `overwolf-extension://{EXTENSION_ID}` as origin.

File: src/app/utils/highlights.ts [hunk 2]
```diff
@@ -1,3 +1,12 @@
+import { Match, MatchHighlight } from '../../types';
+
+const { VITE_PORT } = import.meta.env;
+const baseUrl = `http://localhost:${VITE_PORT}/api`;
```
Comment (lmachens) [inline src/app/utils/highlights.ts]: The server will be deployed with a different hostname, thus, I would add an environment variable for the whole url like `API_URL=https://localhost:3001`.

File: src/app/utils/highlights.ts [hunk 3]
```diff
@@ -83,19 +143,38 @@ export function startCaptureHighlights(): void {
...
   });
 
   overwolf.games.getRunningGameInfo(async (result) => {
     console.log('check if game is already running', result);
     if (result) {
       await getHighlightsAndTurnOn(result.classId);
+
+      if (!activeMatch) {
+        activeMatch = await postNewMatch(result.classId);
+      }
```
Comment (lmachens) [inline src/app/utils/highlights.ts]: Why is it important to check if `activeMatch` is not set?

File: src/app/utils/highlights.ts [hunk 4]
```diff
@@ -122,7 +201,7 @@ type OnProgress = ({ loaded, total }: OnProgressProps) => void;
 
 type UploadToCloudinaryProps = {
   file: Blob;
-  onProgress: OnProgress;
+  onProgress?: OnProgress;
```
Comment (lmachens) [inline src/app/utils/highlights.ts:156]: What's the reason for this change?

File: src/app/utils/highlights.ts [hunk 5]
```diff
@@ -69,6 +78,57 @@ async function getHighlightsAndTurnOn(classId: number) {
...
+    return null;
+  }
+}
+
+async function postNewHighlight({
+  videoSrc,
+  type,
+  timestamp,
+}: MatchHighlight): Promise<ActiveMatch> {
+  if (!activeMatch) return null;
```
Comment (lmachens) [inline src/app/utils/highlights.ts]: I would add `activeMatch` as argument (or `matchId`). This allows to test and use this function in isolation.

**Review by lmachens — COMMENTED** [review]
Good start! 🚀

File: src/app/utils/highlights.ts (hunk 4 above)
Comment (webdevelukas) [reply src/app/utils/highlights.ts:156]: Cause I actually give no onProgress function to the `uploadHighlight` and the `uploadToCloudinary` function.

**Review by lmachens — APPROVED** [review]
🤘

File: src/app/utils/api.ts [hunk 6]
```diff
@@ -15,6 +16,49 @@ export function fetchJSON<T>(
...
+        'Content-Type': 'application/json',
+      },
+      body: data,
+    });
+  } catch (error) {
+    console.log(error);
+    return null;
+  }
+}
+
```
Comment (AlexisRoe) [inline src/app/utils/api.ts]: Would it be useful, to extract the options object to use it in both functions? Propably something like this?
```suggestion

function createFetchOptions (data) {({
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: data,
    }
 )};

export async function postMatch(gameId: number): Promise<Match | null> {
  const { username } = (await getCurrentUser()) || { username: 'unknown' };
  const data = JSON.stringify({ gameId, username });

  try {
    return fetchJSON<Match>('/api/matches', createFetchOptions(data));
  } catch (error) {
    console.log(error);
    return null;
  }
}

```

**Review by AlexisRoe — CHANGES_REQUESTED** [review]
nice start

**Review by AlexisRoe — APPROVED** [review]
gj
